bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/link/ItemChannelLinkRegistry.java

```diff
@@ -53,70 +52,41 @@ public ItemChannelLinkRegistry() {
-        Set<ChannelUID> channelUIDs = new HashSet<>();
-
-        for (ItemChannelLink itemChannelLink : getAll()) {
-            if (itemChannelLink.getItemName().equals(itemName)) {
-                channelUIDs.add(itemChannelLink.getLinkedUID());
-            }
-        }
-
-        return channelUIDs;
+    public Set<ChannelUID> getBoundChannels(final String itemName) {
+        return getLinks(itemName).parallelStream().map(link -> link.getLinkedUID()).collect(Collectors.toSet());
     }
 
     @Override
-    public Set<String> getLinkedItemNames(UID uid) {
-        final Set<String> linkedItems = new LinkedHashSet<>();
-        for (final AbstractLink link : getAll()) {
-            final String itemName = link.getItemName();
-            if (link.getLinkedUID().equals(uid) && itemRegistry.get(itemName) != null) {
-                linkedItems.add(itemName);
-            }
-        }
-        return linkedItems;
+    public Set<String> getLinkedItemNames(final UID uid) {
+        return super.getLinkedItemNames(uid).parallelStream().filter(itemName -> itemRegistry.get(itemName) != null)
+                .collect(Collectors.toSet());
     }
 
-    public Set<Item> getLinkedItems(UID uid) {
-        final Set<Item> linkedItems = new LinkedHashSet<>();
-        for (final AbstractLink link : getAll()) {
-            final String itemName = link.getItemName();
-            Item item = itemRegistry.get(itemName);
-            if (link.getLinkedUID().equals(uid) && item != null) {
-                linkedItems.add(item);
-            }
-        }
-        return linkedItems;
+    public Set<Item> getLinkedItems(final UID uid) {
+        return super.getLinkedItemNames(uid).parallelStream().map(itemName -> itemRegistry.get(itemName))
+                .filter(Objects::nonNull).collect(Collectors.toSet());
     }
 
     /**
      * Returns a set of bound things for the given item name.
      *
      * @param itemName item name
-     * @return set of bound things for the given item name
+     * @return an unmodifiable set of bound things for the given item name
      */
-    public Set<Thing> getBoundThings(String itemName) {
-        Set<Thing> things = new HashSet<>();
-        Collection<ChannelUID> boundChannels = getBoundChannels(itemName);
-
-        for (ChannelUID channelUID : boundChannels) {
-            Thing thing = thingRegistry.get(channelUID.getThingUID());
-            if (thing != null) {
-                things.add(thing);
-            }
-        }
-
-        return things;
+    public Set<Thing> getBoundThings(final String itemName) {
+        return getBoundChannels(itemName).parallelStream()
+                .map(channelUID -> thingRegistry.get(channelUID.getThingUID())).filter(Objects::nonNull)
+                .collect(Collectors.toSet());
     }
 
     @Reference
-    protected void setThingRegistry(ThingRegistry thingRegistry) {
+    protected void setThingRegistry(final ThingRegistry thingRegistry) {
```

Comment:
I think we agreed on not adding "final" for method parameters for readability.

Reply:
See also: https://github.com/eclipse/smarthome/pull/6427#discussion_r229669253
I have added it on existing methods because of a review request.

IIRC we did not agree that "final" must not be used on parameters.
The discussion I remember (I could be wrong) has been about a formatter rule and save action that adds the `final` keyword automatically. You stated that adding it could harm the readability. I accepted to not add that rule but I was not aware of a rule that we must not add final keywords on function parameters.

If I use the following command (and it will not show all usages as it only finds the first parameter (and perhaps some false positives)

```
find -name "*.java" | xargs grep '(final ' | wc -l
```

it already shows that our current code base uses final in more than 600 cases.

And (IMHO) the intention of `final` is not to keep the code more or less readable but to ensure that an argument variable is not reassigned. Shouldn't we more concentrate on safety?